src/get-started/cryptocurrency-advanced.md

```diff
@@ -0,0 +1,801 @@
+pub mod wallet;
+```
+
+## Constants
+
+Let’s define the constants to be used further:
+
+```rust
+/// Unique service ID.
+const CRYPTOCURRENCY_SERVICE_ID: u16 = 128;
+/// Name of the service.
+pub const SERVICE_NAME: &str = "cryptocurrency";
+/// Initial balance of the wallet.
+const INITIAL_BALANCE: u64 = 100;
+```
+
+Implement a service structure and realize a trait for it with all the necessary
+methods and credentials:
+
+```rust
+/// Exonum `Service` implementation.
+#[derive(Default, Debug)]
+pub struct CurrencyService;
+
+impl Service for CurrencyService {
+    // Indicate service name
+    fn service_name(&self) -> &str {
+        SERVICE_NAME
+    }
+
+    // Indicate service ID.
+    fn service_id(&self) -> u16 {
+        CRYPTOCURRENCY_SERVICE_ID
+    }
+
+    // Indicate the way the state hash of the database is calculated.
+    fn state_hash(&self, view: &dyn Snapshot) -> Vec<Hash> {
+        let schema = CurrencySchema::new(view);
+        schema.state_hash()
+    }
+
+    // Interface to create transactions from raw data.
+    fn tx_from_raw(&self, raw: RawTransaction) -> Result<Box<dyn Transaction>, EncodingError> {
+        WalletTransactions::tx_from_raw(raw).map(Into::into)
+    }
+
+    // Indicate public API.
+    fn wire_api(&self, builder: &mut ServiceApiBuilder) {
+        api::CryptocurrencyApi::wire(builder);
+    }
+}
+```
+
+Unlike in previous tutorial where node configuration was realized manually, in
+the advanced tutorial we implement a
+trait that allows us to launch the project with the help of `NodeBuilder`.
+`NodeBuilder` is an abstract that allows to
+easily configure a node for launching. In our case it is responsible for simple
+set up of the network with the
+Cryptocurrency Advanced Service on top of it.
+
+```rust
+#[derive(Debug)]
+pub struct ServiceFactory;
+
+impl fabric::ServiceFactory for ServiceFactory {
+    fn service_name(&self) -> &str {
+        SERVICE_NAME
+    }
+
+    fn make_service(&mut self, _: &Context) -> Box<dyn Service> {
+        Box::new(CurrencyService)
+    }
+}
+```
+
+## Declare Persistent Data
+
+Similarly to a simple cryptocurrency demo we need to declare the data that we
+will store in our blockchain, i.e. the `wallet` type:
+
+```rust
+use exonum::crypto::{Hash, PublicKey};
+encoding_struct! {
+    /// Wallet information stored in the database.
+    struct Wallet {
+        pub_key:            &PublicKey,
+        name:               &str,
+        balance:            u64,
+        history_len:        u64,
+        history_hash:       &Hash,
+    }
+}
+```
+
+As you can see, apart from the already known fields (public key, owner’s name,
+balance), the new wallet type stores
+the **length of the wallet history** and its **hash**. These data are required
+to link the wallet history to other
+histories/data. In this way maps and other indices required for cryptographic
+proofs will be built.
+
+We also realize an auxiliary method for the `Wallet` structure. The method
```

Comment:
&lt;span class="nitpicking"&gt;
Every time I see _realize_ used with 'methods' it throws me off because _implementing_ them seems to be more common expression.
&lt;/span&gt;